### ISSUE TITLE
Making a plugin

### ISSUE THREAD

**Issue by sidchilling** [driftyco/ng-cordova#735]
How to create a new ngCordova plugin for a cordova plugin? I want to make one for https://github.com/vliesaputra/DeviceInformationPlugin

**Comment by aka-Ani** [comment]
I'm not too knowledgeable on how to contribute to the ngCordova project with your custom implementation of a plugin, but making an angular wrapper for a plugin is not too difficult. Basically, take a look at the ng-cordova plugins folder:
https://github.com/driftyco/ng-cordova/tree/master/src/plugins

You can see there is a js file for each wrapper. Let us take the ble.js file for example:
https://github.com/driftyco/ng-cordova/blob/master/src/plugins/ble.js

You can see all you need to do is define a module and factory which implements each of the functions that the plugin provides:
```javascript
angular.module('ngCordova.plugins.ble', [])

  .factory('$cordovaBLE', ['$q', function ($q) {
```
You would have to make your own names and stuff.
Here is the github page for the plugin:
https://github.com/don/cordova-plugin-ble-central

So let's take the scan function. If you were using the plugin directly, the structure is:
`ble.scan(services, seconds, success, failure);`

And in the angular wrapper for this plugin you can see the scan function is implemented as:
```javascript
scan: function (services, seconds) {
        var q = $q.defer();
        ble.scan(services, seconds, function (result) {
          q.resolve(result);
        }, function (error) {
          q.reject(error);
        });
        return q.promise;
      }
```
The principle behind this is called Promises. I'm not an expert on this either, but there are tutorials online that you can read up on. But just by observation, you can see how the success and failure portions become "resolve" and "reject" and you return the "promise". 

You can make these implementations for any/all of the functions you want available from the plugin. And that's kind of the basics of the whole process I think.

**Comment by sidchilling** [comment]
I still could not get this working :(

I am more interested in getting this plugin working for my project rather than contributing to ngCordova. Once, I get this running in my project - I will be in a better position to contribute.

**Comment by aka-Ani** [comment]
Well what is the issue you're running into exactly? Did you make sure to add the plugin to the ionic project with "cordova plugin add ..."? The angular wrappers/ng-cordova still need you to add the plugin separately.

**Comment by sidchilling** [comment]
Here's what I did to get this work in my project -

- Installed the plugin using cordova plugin add
- Did ionic platform rm android
- Did ionic platform add android
- Then I added the following code to ng-cordova.js which I am including in my index.html

    angular.module('ngCordova.plugins.deviceInformation', [])
        .factory('$cordovaDeviceInformation', ['$q', '$window', function ($q, $window) {
            return {
                get: function() {
                    var q = $q.defer();

                    $window.plugins.deviceInformation.get(function (result) {
                        q.resolve(result);
                    }, function(error) {
                        q.reject(error);
                    });
                    return q.promise;
                }
            };
        }]);

**Comment by aka-Ani** [comment]
You got it working? That's great! Honestly though, for such a simple plugin that only has one function, I don't really see the need for constructing an angular wrapper for it. It takes only a couple lines to get the data you need. But in any case, in your code for the wrapper, the code inside the `function()` block after the `defer()` call should basically be the same code as you would have if you were calling the plugin directly without the wrapper. So according to the plugin readme, it would look something like this:
```javascript
var deviceInfo = cordova.require("cordova/plugin/DeviceInformation");
deviceInfo.get(function(result) {
        console.log("result = " + result);
    }, function() {
        console.log("error");
    });
```
But except of console logs, you would put the `resolve` and `reject` functions.

**Comment by sidchilling** [comment]
Hey, I am sorry - I did not get this working.

So, what you are saying is I don't  need an angular plugin and just call the code snippet inside the function that you wrote.

In that case, I get `cordova` is not degine and don't see the cordova/plugin/DeviceInfomation anywhere in  the project? Where should I be looking at?

Regarding the angular wrapper, I tried using it by injecting $cordovaDeviceInformation to my controller but it gives an error that it is undefined.

Any more ideas? I am perfectly ok with not using an angular wrapper.

**Comment by sidchilling** [comment]
I think I got this working - will update in a few moments with the details.

**Comment by xiaoshitou** [comment]
it ok?  i also define my own ngCordova,but it  is undefind in uselly  ,what can i do??

**Comment by kylelib** [comment]
What was your solution? I'm curious how `$window.plugins` is getting insitialized

**Comment by gortok** [comment]
@kylelib  If you're using Angular (which ngCordova depends on), then you simply use the DI syntax to have $window be injected:

```
angular.module('ngCordova.plugins.myPlugin', [])
        .factory('$cordovaMyPlugin', ['$q', '$window', function ($q, $window) {
            return {
                get: function() {
                    var q = $q.defer();
                    $window.plugins.myPlugin.get(function (result) {
                        q.resolve(result);
                    }, function(error) {
                        q.reject(error);
                    });
                    return q.promise;
                }
            };
        }]);
```

**Comment by kylelib** [comment]
@gortok Yep, I totally understand that part. What I'm missing is how `myPlugin` in your example gets attached to `$windows.plugins`. For that matter, how does `plugins` get initialized?

**Comment by ionSpark** [comment]
It doesn't, It get attached to window in the original javascript ($window is just angularjs way of calling "window" so it would be $window.deviceInfo.get

**Comment by kylelib** [comment]
@ionSpark Thanks for your answer. I actually figured it out a few days after posting the question. Turns out this can be controlled via plugin.xml quite easily on the js-module element.

**Comment by ryanmattes** [comment]
How? Please post a link to documentation or an example.

**Comment by kylelib** [comment]
You specify this in the <js-module> node of the plugin.xml for your plugin. It is documented [here ](https://cordova.apache.org/docs/en/latest/plugin_ref/spec.html#js-module)

For example, to attach a plugin named _myPlugin_ defined in _www/myPlugin.js_ to both `$window` and `$window.cordova.plugins.myPlugin` use the following:

```
  <!-- JavaScript interface attached to $window -->
  <js-module src="www/myPlugin.js" name="myPlugin">
    <clobbers target="myPlugin" />
    <clobbers target="cordova.plugins.myPlugin" />
  </js-module>
```
